src/Interpreters/ArithmeticOperationsInAgrFuncOptimize.cpp

```diff
@@ -0,0 +1,286 @@
+#include <Common/typeid_cast.h>
+#include <Parsers/ASTLiteral.h>
+#include <Parsers/ASTFunction.h>
+#include <Interpreters/ArithmeticOperationsInAgrFuncOptimize.h>
+#include <IO/WriteHelpers.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    extern const int UNEXPECTED_AST_STRUCTURE;
+}
+
+bool isConstantField(const Field & field)
+{
+    return field.getType() == Field::Types::Int64 ||
+           field.getType() == Field::Types::UInt64 ||
+           field.getType() == Field::Types::Int128 ||
+           field.getType() == Field::Types::UInt128;
+}
+
+/// scalar values from the first level
+std::pair<ASTs, ASTs> tryGetConst(std::string & name, ASTs & arguments)
+{
+    ASTs const_num;
+    ASTs not_const;
+
+    for (const auto & arg : arguments)
+    {
+        if (const auto * literal = arg->as<ASTLiteral>())
+        {
+            if (isConstantField(literal->value))
+                const_num.push_back(arg);
+            else
+                not_const.push_back(arg);
+        }
+        else
+            not_const.push_back(arg);
+    }
+
+    if ((name == "plus" || name == "multiply") && const_num.size() + not_const.size() != 2)
+    {
+        throw Exception("Wrong number of arguments for function 'plus' or 'multiply' (" + toString(const_num.size() + not_const.size()) + " instead of 2)",
+                        ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+    }
+
+    return {const_num, not_const};
+}
+
+std::pair<ASTs, ASTs> findAllConsts(ASTFunction * func_node, std::string & inter_func_name)
+{
+    if (!func_node->arguments)
+        return {};
+
+    if (!(func_node->arguments->children[0]->as<ASTFunction>()) && (func_node->arguments->children.size() == 2 &&
+        !(func_node->arguments->children[1]->as<ASTFunction>())))
+    {
+        auto arg = func_node->arguments->children;
+        std::pair<ASTs, ASTs> fl = tryGetConst(func_node->name, arg);
+        return fl;
+    }
+    else if ((func_node->arguments->children[0]->as<ASTFunction>() && inter_func_name != func_node->arguments->children[0]->as<ASTFunction>()->name) ||
```

Comment:
Move this expression out as bool variable and add informative name to it. What does this expression mean?